### PR TITLE
feat: Add warning for submissions for requests that cannot currently be fulfilled [DET-6410]

### PR DESCRIPTION
**Pull request by julian-determined-ai**
## Description

feat: Add warning for submissions for requests that cannot currently be fulfilled [DET-6410]

Presently, if a user requests a job with a larger number of slots than are currently available the job may be stuck in queue indefinitely. This PR shows a **warning** message in the CLI and WebUI to a user letting them know that too many slots may have been requested. 

**important note:** Since we cannot accurately tell if more slots may become available in the future we are **not** blocking job submission, only showing a warning to the user.

Screen Shot 2022-11-03 at 10 04 00 AM

Screen Shot 2022-11-03 at 10 03 26 AM

Screen Shot 2022-11-03 at 10 33 32 AM

## Test Plan

1. Attempt to create a notebook in the WebUI requesting more slots than available. Ensure that you see a warning on the about the number of slots. 
2. Attempt to create a notebook in the CLI requesting more slots than available. Ensure that you see a warning about the number of slots. 
3. Attempt to fork an experiment and in the WebUI and in the configuration specify more slots than available. Ensure that you see a warning about the number of slots. 
4. Create an experiment in the CLI that requires more slots than available. Ensure that you see a warning about the number of slots. 
5. Repeat steps 1-4 instead specify a number a slots equal to or less than the amount available and ensure that you do not see any warning messages. 

## Checklist
- [x] Changes have been manually QA'd
~- [ ] User-facing API changes need the "User-facing API Change" label.~
~- [ ] Release notes should be added as a separate file under `docs/release-notes/`.~
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
~- [ ] Licenses should be included for new code which was copied and/or modified from any external code.~

[DET-6410]: https://determinedai.atlassian.net/browse/DET-6410?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: examples/computer_vision/cifar10_pytorch/const.yaml [hunk 1]
```diff
@@ -6,12 +6,12 @@ hyperparameters:
...
   layer3_dropout: 0.5
   global_batch_size: 32
-records_per_epoch: 50000
+records_per_epoch: 5
 searcher:
   name: single
   metric: validation_error
   max_length:
-    epochs: 32
+    epochs: 2
```
[the remark this answers is not in the document]
Comment (julian-determined-ai) [reply examples/computer_vision/cifar10_pytorch/const.yaml:14]: I do this all the time ha sorry let me remove.

File: master/internal/rm/agent_resource_manager.go [hunk 2]
```diff
@@ -95,45 +101,83 @@ func (a AgentResourceManager) IsReattachEnabledForRP(ctx actor.Messenger, rpName
...
 		}
-		return resp.PoolName, nil
+		rp.Name = resp.PoolName
+		rp.CurrentMaxSlotsExceeded = currentMaxSlotsExceeded
+		return rp, nil
 	}
 
 	if err := a.ValidateResourcePool(ctx, name); err != nil {
-		return "", fmt.Errorf("validating pool: %w", err)
+		return rp, fmt.Errorf("validating pool: %w", err)
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:158]: can you make the respond a pointer/default object, so we don't return something that looks half usable?

**Review by stoksc — COMMENTED** [review]
looks like good progress. maybe we should chat offline about some of this stuff?

**Comment by mapmeld** [comment]
Is there a reason that the experiment resources: `resources.slots_per_trial` check works in WebUI / forking, but isn't reported when creating an experiment from the CLI?

**Comment by julian-determined-ai** [comment]
> Is there a reason that the experiment resources: `resources.slots_per_trial` check works in WebUI / forking, but isn't reported when creating an experiment from the CLI?

Yes the reason is because I didn't add that warning to the CLI 😅 I will add it so that the CLI And WebUI have identical behavior

File: harness/determined/cli/notebook.py [hunk 3]
```diff
@@ -28,7 +28,11 @@ def start_notebook(args: Namespace) -> None:
         return
 
     nb = resp.notebook
-    maxCurrentSlotsExceeded = True if resp.maxCurrentSlotsExceeded else False
+    warnings = resp.warnings
+
+    maxCurrentSlotsExceeded = (
+        warnings and v1LaunchWarning.LAUNCH_WARNING_MAX_CURRENT_SLOTS_EXCEEDED in warnings
```
Comment (stoksc) [inline harness/determined/cli/notebook.py:34]: max current slots is a misleading way to coin this, i think just "CURRENT_SLOTS_EXCEEDED" does a better job implying this may be temporary.

File: master/internal/rm/agent_resource_manager.go [hunk 4]
```diff
@@ -95,16 +96,38 @@ func (a AgentResourceManager) IsReattachEnabledForRP(ctx actor.Messenger, rpName
...
-	ctx actor.Messenger, name string, slots int, command bool,
+	ctx actor.Messenger, name string, slots int,
 ) (string, error) {
 	// If the resource pool isn't set, fill in the default at creation time.
 	if name == "" && slots == 0 {
 		req := sproto.GetDefaultAuxResourcePoolRequest{}
 		resp, err := a.GetDefaultAuxResourcePool(ctx, req)
 		if err != nil {
-			return "", fmt.Errorf("defaulting to aux pool: %w", err)
+			return resp.PoolName, fmt.Errorf("defaulting to aux pool: %w", err)
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:130]: why this change? without looking, i would i think in the event of an an error, checking `resp.PoolName` may be a nil pointer panic.

File: master/internal/rm/agent_resource_manager.go [hunk 5]
```diff
@@ -113,27 +136,54 @@ func (a AgentResourceManager) ResolveResourcePool(
 		req := sproto.GetDefaultComputeResourcePoolRequest{}
 		resp, err := a.GetDefaultComputeResourcePool(ctx, req)
 		if err != nil {
-			return "", fmt.Errorf("defaulting to compute pool: %w", err)
+			return name, fmt.Errorf("defaulting to compute pool: %w", err)
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:139]: same q

File: master/internal/rm/agent_resource_manager.go [hunk 6]
```diff
@@ -113,27 +136,54 @@ func (a AgentResourceManager) ResolveResourcePool(
...
 		}
 		return resp.PoolName, nil
 	}
 
 	if err := a.ValidateResourcePool(ctx, name); err != nil {
-		return "", fmt.Errorf("validating pool: %w", err)
+		return name, fmt.Errorf("validating pool: %w", err)
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:145]: same q again

File: master/internal/rm/agent_resource_manager.go [hunk 7]
```diff
@@ -113,27 +136,54 @@ func (a AgentResourceManager) ResolveResourcePool(
...
-			return "", errors.New("request unfulfillable, please try requesting less slots")
+			return errors.New("request unfulfillable, please try requesting less slots")
 		}
 	}
-	return name, nil
+	return nil
+}
+
+// GetResourcePoolAvailability is a default implementation to satisfy the interface.
+func (a AgentResourceManager) GetResourcePoolAvailability(ctx actor.Messenger,
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:170]: maybe validate?
```suggestion
func (a AgentResourceManager) ValidateResourcePoolAvailability(ctx actor.Messenger,
```

File: master/internal/rm/agent_resource_manager.go [hunk 8]
```diff
@@ -113,27 +136,54 @@ func (a AgentResourceManager) ResolveResourcePool(
...
+	name string, slots int) (
+	[]command.LaunchWarning,
+	error,
+) {
+	launchWarnings := []command.LaunchWarning{}
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:175]:
```suggestion
	var warnings []command.LaunchWarning
```

File: master/internal/rm/agent_resource_manager.go [hunk 9]
```diff
@@ -113,27 +136,54 @@ func (a AgentResourceManager) ResolveResourcePool(
...
+	if slots == 0 {
+		return launchWarnings, nil
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:177]: maybe just this, again? not going to comment this anymore.. i think i've highlighted the pattern. if im missing something lmk.
```suggestion
		return nil, nil
```

File: master/internal/rm/agent_resource_manager.go (hunk 4 above)
Comment (stoksc) [reply master/internal/rm/agent_resource_manager.go:130]: the reason i'm pushing returning explicit nil or "" - not returning anything meaningful when you error - is that functions/apis that allow partial success are easy to misuse.

File: master/internal/rm/kubernetes_resource_manager.go [hunk 10]
```diff
@@ -96,6 +105,22 @@ func (k KubernetesResourceManager) ValidateResourcePool(ctx actor.Messenger, nam
...
+func (k KubernetesResourceManager) GetResourcePoolAvailability(
+	ctx actor.Messenger,
+	name string,
+	slots int,
+) ([]command.LaunchWarning, error) {
+	launchWarnings := []command.LaunchWarning{}
+	if slots == 0 {
+		return launchWarnings, nil
+	}
+	if name != "" && name != KubernetesDummyResourcePool {
```
Comment (stoksc) [inline master/internal/rm/kubernetes_resource_manager.go:118]: do we need to duplicate the `ValidateResourcePool` check for some reason. I would think this should always return `return nil, nil` until we implement this.

File: harness/determined/common/api/request.py [hunk 11]
```diff
@@ -56,13 +61,14 @@ def make_interactive_task_url(
...
     )
-    task_web_url = "{}/interactive/{}/{}/{}/{}/{}".format(
+    task_web_url = "{}/interactive/{}/{}/{}/{}/{}/{}".format(
         public_url,
         task_id,
         task_type,
         parse.quote(description),
         resource_pool,
         parse.quote_plus(wait_page_url),
+        str(currentSlotsExceeded).lower(),
```
[the remark this answers is not in the document]
Comment (julian-determined-ai) [reply harness/determined/common/api/request.py]: updated this.

File: e2e_tests/tests/command/command.py [hunk 12]
```diff
@@ -24,8 +24,10 @@ def __init__(self, process: subprocess.Popen, detach: bool = False):
         else:
             iterator = iter(self.process.stdout)  # type: ignore
             line = next(iterator)
-            m = re.search(rb"Scheduling .* \(id: (.*)\)", line)
-            assert m is not None
+            m = None
+            while line and not m:
```
[the remark this answers is not in the document]
Comment (julian-determined-ai) [reply e2e_tests/tests/command/command.py:28]: Ultimately no, with the updates here if all lines are exhausted and `m` is still now we will return `None` and not a valid task_id so all tests that would have failed previously will still fail now.

File: master/internal/rm/kubernetes_resource_manager.go (hunk 10 above)
Comment (julian-determined-ai) [reply master/internal/rm/kubernetes_resource_manager.go:114]: updated

File: e2e_tests/tests/command/command.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (stoksc) [reply e2e_tests/tests/command/command.py:28]: yeah, but, shouldn't we explicitly fail like we used to instead of relying on something downstream to fail because task_id is None, or am I missing something?

[the remark this answers is not in the document]
Comment (julian-determined-ai) [reply e2e_tests/tests/command/command.py:28]: Sorry I was confused,  I missed that I removed the assertion it has been re added.

File: examples/nas/gaea_pytorch/search/const.yaml [hunk 13]
```diff
@@ -25,7 +25,7 @@ hyperparameters:
     nodes: 4
 
 resources:
-  slots_per_trial: 2
+  slots_per_trial: 6
```
[the remark this answers is not in the document]
Comment (julian-determined-ai) [reply examples/nas/gaea_pytorch/search/const.yaml]: fixed

File: e2e_tests/tests/test_sdk.py [hunk 14]
```diff
@@ -28,10 +28,10 @@ def test_completed_experiment_and_checkpoint_apis(client: _client.Determined) ->
     emptydir = tempfile.mkdtemp()
     try:
         model_def = conf.fixtures_path("no_op/model_def.py")
-        exp = client.create_experiment(config, emptydir, includes=[model_def])
+        exp, _ = client.create_experiment(config, emptydir, includes=[model_def])
```
[the remark this answers is not in the document]
Comment (julian-determined-ai) [reply e2e_tests/tests/test_sdk.py:31]: If this is too big of a change I am happy to revert the changes in the CLI

File: harness/determined/cli/tensorboard.py [hunk 15]
```diff
@@ -29,12 +29,19 @@ def start_tensorboard(args: Namespace) -> None:
 
     req_body["files"] = context.read_legacy_context(args.context, args.include)
 
-    resp = api.post(args.master, "api/v1/tensorboards", json=req_body).json()["tensorboard"]
+    api_resp = api.post(args.master, "api/v1/tensorboards", json=req_body).json()
```
Comment (rb-determined-ai) [inline harness/determined/cli/tensorboard.py:32]: if you are going to touch an `api.post()`, please update it to use the generated bindings.

File: e2e_tests/tests/test_sdk.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply e2e_tests/tests/test_sdk.py:31]: the `client` module is the Python SDK, so yeah we shouldn't be changing it.

Why not just do `logging.warning()` in `create_experiment()`?

File: harness/determined/common/experimental/determined.py [hunk 16]
```diff
@@ -129,7 +129,7 @@ def create_experiment(
         config: Union[str, pathlib.Path, Dict],
         model_dir: Union[str, pathlib.Path],
         includes: Optional[Iterable[Union[str, pathlib.Path]]] = None,
-    ) -> experiment.ExperimentReference:
+    ) -> Tuple[experiment.ExperimentReference, Optional[Sequence[bindings.v1LaunchWarning]]]:
```
Comment (rb-determined-ai) [inline harness/determined/common/experimental/determined.py:132]: This is a public API, don't change it

File: harness/determined/experimental/client.py [hunk 17]
```diff
@@ -145,7 +145,7 @@ def create_experiment(
     config: Union[str, pathlib.Path, Dict],
     model_dir: str,
     includes: Optional[Iterable[Union[str, pathlib.Path]]] = None,
-) -> ExperimentReference:
+) -> Tuple[ExperimentReference, Optional[Sequence[bindings.v1LaunchWarning]]]:
```
Comment (rb-determined-ai) [inline harness/determined/experimental/client.py:148]: ditto

File: harness/determined/common/api/experiment.py [hunk 18]
```diff
@@ -127,7 +128,10 @@ def create_experiment(
     archived: bool = False,
     activate: bool = True,
     additional_body_fields: Optional[Dict[str, Any]] = None,
-) -> int:
+) -> Tuple[int, Optional[Sequence[bindings.v1LaunchWarning]]]:
```
Comment (rb-determined-ai) [inline harness/determined/common/api/experiment.py:131]: technically, this is an internal function so you can change this API if you'd like to.

Somewhere I have some work to [clean this up](https://github.com/determined-ai/determined/pull/4914).  So hopefully this function just evaporates eventually.  Not sure that affects what you to this function today though.

File: harness/determined/cli/tensorboard.py [hunk 19]
```diff
@@ -56,9 +63,9 @@ def start_tensorboard(args: Namespace) -> None:
                             resource_pool=resp["resourcePool"],
                             description=resp["description"],
                             task_type="tensorboard",
+                            currentSlotsExceeded=currentSlotsExceeded,
                         ),
                     )
-
```
Comment (rb-determined-ai) [inline harness/determined/cli/tensorboard.py:61]: please avoid extraneous whitespace changes

File: harness/determined/common/api/request.py [hunk 20]
```diff
@@ -356,3 +363,37 @@ def ws(host: str, path: str) -> WebSocket:
     token = authentication.must_cli_auth().get_session_token()
     websocket.add_header("Authorization".encode(), "Bearer {}".format(token).encode())
     return WebSocket(websocket)
+
+
+def parse_warnings(
```
Comment (rb-determined-ai) [inline harness/determined/common/api/request.py:368]: Pretty sure this function is not necessary.

Pretty sure `bindings.v1LaunchWarning(x)` is what you are looking for; it should work with whatever the REST API emits.

This is a snippet from the generated bindings for how the conversion is done there:

```python
[v1LaunchWarning(x) for x in obj["warnings"]] if obj["warnings"] is not None else None
```

File: harness/determined/common/api/request.py [hunk 21]
```diff
@@ -44,7 +45,12 @@ def maybe_upgrade_ws_scheme(master_address: str) -> str:
 
 
 def make_interactive_task_url(
-    task_id: str, service_address: str, description: str, resource_pool: str, task_type: str
+    task_id: str,
+    service_address: str,
+    description: str,
+    resource_pool: str,
+    task_type: str,
+    currentSlotsExceeded: bool,
```
Comment (rb-determined-ai) [inline harness/determined/common/api/request.py:53]: Out of curiosity, why is this value passed through the client, and not through the master?

File: harness/determined/common/api/request.py [hunk 22]
```diff
@@ -356,3 +363,37 @@ def ws(host: str, path: str) -> WebSocket:
...
+
+
+def parse_warnings(
+    warnings: Optional[Union[Sequence[bindings.v1LaunchWarning], Sequence[int]]]
+) -> Optional[Sequence[bindings.v1LaunchWarning]]:
+
+    # Warnings may be passed in as either an enum value or integer value
+    warning_message_int_to_enum_value = {
+        1: bindings.v1LaunchWarning.LAUNCH_WARNING_CURRENT_SLOTS_EXCEEDED
+    }
```
Comment (rb-determined-ai) [inline harness/determined/common/api/request.py:375]: When would this happen?

File: harness/determined/common/api/request.py [hunk 23]
```diff
@@ -356,3 +363,37 @@ def ws(host: str, path: str) -> WebSocket:
...
+def handle_warnings(warnings: Optional[Sequence[bindings.v1LaunchWarning]]) -> None:
+    warning_message_map = {
+        bindings.v1LaunchWarning.LAUNCH_WARNING_CURRENT_SLOTS_EXCEEDED: (
+            "Warning: The requested job requires more slots than currently available. "
+            "You may need to increase cluster resources in order for the job to run."
+        )
+    }
+    if warnings:
+        for warning in warnings:
+            print(colored(warning_message_map[warning], "yellow"))
```
Comment (rb-determined-ai) [inline harness/determined/common/api/request.py:399]: Generally, you don't want to print warnings to `stdout` in a CLI because `stdout` is commonly used in bash pipelines.  You don't want warnings to break those pipelines, and you do want the warnings to show even when a pipeline is in use.

Easy fix though, just print to `file=sys.stderr` instead.

File: harness/determined/common/api/request.py [hunk 24]
```diff
@@ -356,3 +363,37 @@ def ws(host: str, path: str) -> WebSocket:
...
+    return warning_enum_values
+
+
+def handle_warnings(warnings: Optional[Sequence[bindings.v1LaunchWarning]]) -> None:
+    warning_message_map = {
+        bindings.v1LaunchWarning.LAUNCH_WARNING_CURRENT_SLOTS_EXCEEDED: (
+            "Warning: The requested job requires more slots than currently available. "
+            "You may need to increase cluster resources in order for the job to run."
+        )
+    }
```
Comment (rb-determined-ai) [inline harness/determined/common/api/request.py:396]: Can you make this map a constant in `common/api/_util.py`, then expose it in the `determined.common.api` namespace by importing it into the `common/api/__init__.py` file?  (see the pattern in that file).  This constant belongs there because it is useful to both CLI and non-CLI code.

Then can you separate the printing into a CLI-specific function.  Put that in `cli/_util.py`, then expose it into the `determined.cli` namespace by importing it in the `cli/__init__.py` file.  This particular function is only useful to CLI functions because `print()` shouldn't be called by library code (so, nothing in `determined.experimental.client`, for instance).  But the CLI is an application, not a library, and printing yellow errors is totally fine there.

File: harness/determined/common/api/experiment.py [hunk 25]
```diff
@@ -147,7 +151,15 @@ def create_experiment(
...
     if validate_only:
-        return 0
+        return 0, None
+
+    try:
+        warnings = r.json().get("warnings")
+        if warnings:
+            launch_warnings = list(bindings.v1LaunchWarning)
+            warnings = [launch_warnings[warning].value for warning in warnings]
+    except json.decoder.JSONDecodeError:
```
Comment (rb-determined-ai) [inline harness/determined/common/api/experiment.py:161]: A few things here:

- When you use a try/except, make your try clause as small as possible.  Put the other stuff in an `else` statement.  Basically, you want to be clear about what code might raise an exception:
  ```python
  try:
    j = r.json()
  except json.decoder.JSONDecodeError:
    warnings = None  # here instead of the top of the function
  else:
    warnings = j.get("warnings")
    if warnings is not None:
      warnings = [bindings.v1LaunchWarning(w) for w in warnings]
  ```

- In this particular case, why even bother with the `r.json()` failure?  We'll see other stack traces when that happens anyway (`req.post()` would fail).

File: harness/determined/cli/tensorboard.py (hunk 15 above)
Comment (julian-determined-ai) [reply harness/determined/cli/tensorboard.py]: updated

File: harness/determined/common/experimental/determined.py (hunk 16 above)
Comment (julian-determined-ai) [reply harness/determined/common/experimental/determined.py]: reverted these changes

File: harness/determined/cli/tensorboard.py (hunk 19 above)
Comment (julian-determined-ai) [reply harness/determined/cli/tensorboard.py:61]: reverted this

File: harness/determined/common/api/request.py (hunk 22 above)
Comment (julian-determined-ai) [reply harness/determined/common/api/request.py]: removed this. Relates to: https://github.com/determined-ai/determined/pull/5376#discussion_r1028420540

File: harness/determined/common/api/request.py (hunk 23 above)
Comment (julian-determined-ai) [reply harness/determined/common/api/request.py]: updated

File: harness/determined/common/api/request.py (hunk 24 above)
Comment (julian-determined-ai) [reply harness/determined/common/api/request.py]: updated

File: harness/determined/common/api/experiment.py (hunk 25 above)
Comment (julian-determined-ai) [reply harness/determined/common/api/experiment.py:156]: updated

File: harness/determined/common/api/request.py (hunk 21 above)
Comment (julian-determined-ai) [reply harness/determined/common/api/request.py:55]: could you elaborate on why you think this value should be passed through the master?

File: master/internal/api_command.go [hunk 26]
```diff
@@ -52,27 +53,30 @@ type protoCommandParams struct {
 }
 
 func (a *apiServer) getCommandLaunchParams(ctx context.Context, req *protoCommandParams) (
-	*tasks.GenericCommandSpec, error,
+	*tasks.GenericCommandSpec, []command.LaunchWarning, error,
 ) {
 	var err error
 
+	var launchWarnings []command.LaunchWarning
```
Comment (stoksc) [inline master/internal/api_command.go:60]: don't think you need this decl anymore (change the set to `:=`)

File: master/internal/experiment.go [hunk 27]
```diff
@@ -119,18 +120,28 @@ type (
 // and log. If the input object has no ID set, also create a new experiment in the database and set
 // the returned object's ID appropriately.
 func newExperiment(m *Master, expModel *model.Experiment, taskSpec *tasks.TaskSpec) (
-	*experiment, error,
+	*experiment, []command.LaunchWarning, error,
 ) {
 	conf := &expModel.Config
-
+	var launchWarnings []command.LaunchWarning
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/experiment.go:126]: same here, again (generally better to simultaneously create and properly initialize things)

File: master/internal/rm/actor_resource_manager.go [hunk 28]
```diff
@@ -36,13 +37,36 @@ func (r *ActorResourceManager) ResolveResourcePool(
...
 ) (string, error) {
 	return name, nil
 }
 
+// ValidateResources is a default implementation to satisfy the interface, mostly for tests.
+func (r *ActorResourceManager) ValidateResources(
+	ctx actor.Messenger,
+	name string,
+	slots int,
+	command bool,
```
Comment (stoksc) [inline master/internal/rm/actor_resource_manager.go:49]: can we remove this and collapse `ValidateResources` and `ValidateCommandResources` into one?

File: master/internal/rm/agent_resource_manager.go [hunk 29]
```diff
@@ -95,9 +96,31 @@ func (a AgentResourceManager) IsReattachEnabledForRP(ctx actor.Messenger, rpName
 	return false
 }
 
+// CheckMaxSlotsExceeded checks if the job exceeded the maximum number of slots.
+func (a AgentResourceManager) CheckMaxSlotsExceeded(
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:100]: one thing i can see now; on every new saas cluster, someone will see this on first job submission, every time. maybe we should make it seem less severe if we know the cluster can autoscale?

File: master/internal/rm/agent_resource_manager.go [hunk 30]
```diff
@@ -121,19 +144,47 @@ func (a AgentResourceManager) ResolveResourcePool(
...
+		return nil, nil
+	}
+	currentMaxSlotsExceeded, err := a.CheckMaxSlotsExceeded(ctx, name, slots)
+	if err != nil {
+		return nil, fmt.Errorf("validating request for (%s, %d): %w", name, slots, err)
+	}
+	if currentMaxSlotsExceeded {
+		launchWarnings = append(launchWarnings, command.CurrentSlotsExceeded)
+	}
+	return launchWarnings, nil
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:187]: use a switch here

File: master/internal/rm/agent_resource_manager.go [hunk 31]
```diff
@@ -121,19 +144,47 @@ func (a AgentResourceManager) ResolveResourcePool(
...
+	return nil
+}
+
+// ValidateResourcePoolAvailability is a default implementation to satisfy the interface.
+func (a AgentResourceManager) ValidateResourcePoolAvailability(ctx actor.Messenger,
+	name string, slots int) (
+	[]command.LaunchWarning,
+	error,
+) {
+	var launchWarnings []command.LaunchWarning
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:175]: declare this closer to where it is used

File: proto/src/determined/command/v1/command.proto [hunk 32]
```diff
@@ -46,4 +46,4 @@ message Command {
   string exit_status = 12;
   // The associated job id.
   string job_id = 13;
-}
+}
```
Comment (stoksc) [inline proto/src/determined/command/v1/command.proto:49]:
```suggestion
}

```

File: master/internal/rm/actor_resource_manager.go (hunk 28 above)
Comment (julian-determined-ai) [reply master/internal/rm/actor_resource_manager.go:49]: Depends seems like this would involve updating everywhere that we check the relevant `sproto` i.e https://github.com/determined-ai/determined/blob/7832a93ef7666086cd9a2b185161dde498a07401/master/internal/rm/kubernetes_resource_manager.go#L253

If that sounds right and you think that refactor is worth it I can work on those updates. Also if you had some other type of refactor in mind let me know.

File: master/internal/rm/agent_resource_manager.go (hunk 29 above)
Comment (julian-determined-ai) [reply master/internal/rm/agent_resource_manager.go:100]: Good point, do we have a way of being able to tell if the cluster can autoscale?

File: master/internal/rm/agent_resource_manager.go [hunk 33]
```diff
@@ -172,19 +172,20 @@ func (a AgentResourceManager) ValidateResourcePoolAvailability(ctx actor.Messeng
...
 		return nil, nil
 	}
 	currentMaxSlotsExceeded, err := a.CheckMaxSlotsExceeded(ctx, name, slots)
 	if err != nil {
 		return nil, fmt.Errorf("validating request for (%s, %d): %w", name, slots, err)
 	}
-	if currentMaxSlotsExceeded {
-		launchWarnings = append(launchWarnings, command.CurrentSlotsExceeded)
+
+	switch currentMaxSlotsExceeded {
```
Comment (stoksc) [inline master/internal/rm/agent_resource_manager.go:183]: the whole thing, e.g.
```
switch exceeded, err := a.CheckMaxSlotsExceeded(ctx, name, slots); {
case err != nil:
    ...
case exceeded:
    ...
default:
    ....
```

File: master/internal/rm/agent_resource_manager.go (hunk 29 above)
Comment (stoksc) [reply master/internal/rm/agent_resource_manager.go:100]: yeah, we should be able to tell if it is a static or dynamic cluster.

Comment (julian-determined-ai) [reply master/internal/rm/agent_resource_manager.go:100]: any pointers there?

File: master/internal/sproto/resource_pool.go [hunk 34]
```diff
@@ -0,0 +1,13 @@
+package sproto
+
+type (
+	// CapacityCheck checks the potential available slots in a resource pool.
+	CapacityCheck struct {
+		Slots int
+	}
+	// CapacityCheckResponse is the response to a CapacityCheck message.
+	CapacityCheckResponse struct {
+		SlotsAvailable   int
```
Comment (julian-determined-ai) [inline master/internal/sproto/resource_pool.go:10]: The `slotsAvailable` is not being used but I thought it would be great information to have for future use.

**Review by stoksc — COMMENTED** [review]
lgtm, just a few final comments

**Review by stoksc — APPROVED** [review]
